### PR TITLE
Rock Anomaly

### PR DESCRIPTION
**Pull request by brainfood1183**
Added - rock anomaly. The rock anomaly spawns asteroid walls within its radius as well as crystals.
Added - new type of asteroid which contains ore crab mobs. ore crabs are elemental mobs that when destroyed spawn ore, they can be found within asteroids spawned by the rock anomaly and there is a new asteroid type specifically for spawning them that can be used for expeditions etc.
Added- random crystal spawner.
Added - quartz crab elemental mob.
Added - iron (steel) crab elemental mob.
Added - Uranium crab elemental mob.

untitled-320

https://github.com/space-wizards/space-station-14/assets/113240905/bd057954-c799-459d-84d1-7ac405ca3394

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl:
- add: Rock anomalies have been detected by nanotrasen scientists.

### CODE REVIEW

**Comment by deltanedas** [comment]
>  they can be found when mining high yield asteroids (0.3)

this is bad since it would make asteroids pve which were previously just pure mining

**Comment by NancokPS2** [comment]
What if they where reserved for wreckage layouts like an abandoned mining platform?  
  
Since wreckage can have mobs

**Comment by CrigCrag** [comment]
> > they can be found when mining high yield asteroids (0.3)
> 
> this is bad since it would make asteroids pve which were previously just pure mining

why should asteroids randomly pulled in from the depths of space be completely safe

**Comment by deltanedas** [comment]
> > > they can be found when mining high yield asteroids (0.3)
> > 
> > 
> > this is bad since it would make asteroids pve which were previously just pure mining
> 
> why should asteroids randomly pulled in from the depths of space be completely safe

https://github.com/space-wizards/space-station-14/issues/16970

**Comment by Aisu9** [comment]
now we just need to have the glowing crystal actually dropping crystal on destruction as intended

**Comment by brainfood1183** [comment]
> > they can be found when mining high yield asteroids (0.3)
> 
> this is bad since it would make asteroids pve which were previously just pure mining

yea this has been changed, now also it is distinctly clear which asteroids may contain orecrab (will be multiple types not just quartz)

untitled-322

**Comment by Aisu9** [comment]
maybe i should ask instead of doing what i did, are glowing crystals meant to drop glowing chunks when mined/destroyed? i remember seeing items and sprites for "smaller" version but actually never managed to extract anything from them

**Comment by brainfood1183** [comment]
> maybe i should ask instead of doing what i did, are glowing crystals meant to drop glowing chunks when mined/destroyed? i remember seeing items and sprites for "smaller" version but actually never managed to extract anything from them

not really sure what they are supposed to do other than look pretty tbh

**Comment by Anzarot121** [comment]
Isn't this anomaly too safe compared to the others?

**Comment by mirrorcult** [comment]
doesnt necessarily need to be explicitly dangerous, just needs to be interesting

**Comment by brainfood1183** [comment]
> Isn't this anomaly too safe compared to the others?

It is fairly safe compared to the others but we have seen issues with some of the more dangerous anomalies always being decayed due to the risk.

**Comment by EmoGarbage404** [comment]
Hear me out: gibtonite

**Comment by deltanedas** [comment]
:trollface: salv new validhunting powerhouse with 1ms gibtonite

**Comment by JoeHammad1844** [comment]
image
is that michael from fath: the unholy trinity?

**Comment by EmoGarbage404** [comment]
I like how this is conceptually. The little rock monsters are really good. I really only have 2 design concerns:
1. Is the rock generation too easy to manage? From the video, it looked like it could be dealt with by a group of 1-2 people.
2. Is there anything more you could do with the supercritical? Spawning a ton of rocks is fine but maybe it would be cool if it would replace windows/walls nearby with rocks or even spawn a horde of those ore crabs.

**Comment by brainfood1183** [comment]
> I like how this is conceptually. The little rock monsters are really good. I really only have 2 design concerns:
> 
>     1. Is the rock generation too easy to manage? From the video, it looked like it could be dealt with by a group of 1-2 people.
> 
>     2. Is there anything more you could do with the supercritical? Spawning a ton of rocks is fine but maybe it would be cool if it would replace windows/walls nearby with rocks or even spawn a horde of those ore crabs.

I thought about maybe a rock version of kudzu that would spread asteroid rocks everywhere, but i think it will be fine without, granted it could be managed with 2 people fairly well if they have pickaxes but chances are they will have to find the anomaly, realize its a rock anomaly, then go to cargo or request for a salv to come help with anomaly, by time they have a pickaxe it could have triggered multiple times. Thats if there is a spare one in cargo.

File: Resources/Prototypes/Entities/Markers/Spawners/Random/crystal.yml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+          state: crystal_cyan
+    - type: RandomSpawner
+      prototypes:
+        - CrystalGreen
+        - CrystalPink
+        - CrystalOrange
+        - CrystalBlue
+        - CrystalCyan
+      chance: 0.7
+      offset: 0.4
```
Comment (TheShuEd) [inline Resources/Prototypes/Entities/Markers/Spawners/Random/crystal.yml:19]: Offset doesn't make sense, the crystals will spawn in the middle of the tile

File: Resources/Prototypes/Entities/Mobs/NPCs/elemental.yml [hunk 2]
```diff
@@ -0,0 +1,208 @@
+- type: entity
+  save: false
+  abstract: true
+  id: MobElementalBase
```
Comment (EmoGarbage404) [inline Resources/Prototypes/Entities/Mobs/NPCs/elemental.yml:4]: you can get rid of like half of these if you use composition to inherit from the base mob prototypes. Look at how BaesSimpleMob does it.

File: Resources/Prototypes/Entities/Mobs/NPCs/elemental.yml [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+    context: "human"
+  - type: InputMover
+  - type: MobMover
+  - type: ZombieImmune
+
+- type: entity
+  parent: MobElementalBase
+  id: MobQuartzCrab
+  name: quartz crab
+  description:
```
Comment (EmoGarbage404) [inline Resources/Prototypes/Entities/Mobs/NPCs/elemental.yml:82]: all of these have empty descriptions. If you don't want to do custom ones for each mob, just remove these empty fields and put something generic on the base mob.

File: Resources/Prototypes/Entities/Markers/Spawners/Random/crystal.yml [hunk 4]
```diff
@@ -0,0 +1,19 @@
+- type: entity
+  name: Crystal Spawner
+  id: CrystalSpawner
+  parent: MarkerBase
+  components:
```
Comment (EmoGarbage404) [inline Resources/Prototypes/Entities/Markers/Spawners/Random/crystal.yml:5]: put a suffix `[70]` to indicate that this has a random chance of 70.

File: Resources/Prototypes/Entities/Structures/Walls/asteroid.yml [hunk 5]
```diff
@@ -17,6 +17,10 @@
     mode: NoSprite
   - type: SmoothEdge
   - type: Sprite
+    variants: 2
+    placementVariants:
+    - 1.0
+    - 1.0
```
Comment (EmoGarbage404) [inline Resources/Prototypes/Entities/Structures/Walls/asteroid.yml:23]: you put this on spritecomp when i think you meant to put it on some kind of floor tile. Sprites don't support anything like this (i'm surprised the linter isn't mad)

File: Resources/Prototypes/Entities/Mobs/NPCs/elemental.yml [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+    - DoorBumpOpener
+    - ShoesRequiredStepTriggerImmune
+  - type: MobState
+    allowedStates:
+      - Alive
+      - Dead
+  - type: MobThresholds
+    thresholds:
+      0: Alive
+      120: Dead
```
Comment (EmoGarbage404) [inline Resources/Prototypes/Entities/Mobs/NPCs/elemental.yml:56]: all of these guys could serve to be significantly stronger + do more damage. As it stands, they are the main active threat of the anomaly and provide a fair amount of materials. They should be able to kill an unprepared tider.

**Review by EmoGarbage404 — CHANGES_REQUESTED** [review]
Really solid concept. Mostly just some mild cleanup and balancing to look at.

**Comment by notquitehadouken** [comment]
pros: pretty colors
cons: if you are close enough to see the pretty colors, you are dead

File: Resources/Prototypes/Entities/Mobs/NPCs/elemental.yml (hunk 2 above)
Comment (brainfood1183) [reply Resources/Prototypes/Entities/Mobs/NPCs/elemental.yml:4]: wouldnt i also inherit like blood solution etc?

File: Resources/Prototypes/Entities/Structures/Walls/asteroid.yml (hunk 5 above)
Comment (brainfood1183) [reply Resources/Prototypes/Entities/Structures/Walls/asteroid.yml:23]: lol, yea i attempted to do this but it didnt work so i created a spawner with 2 entities and forgot to remove it again.

File: Resources/Prototypes/Entities/Mobs/NPCs/elemental.yml (hunk 6 above)
Comment (brainfood1183) [reply Resources/Prototypes/Entities/Mobs/NPCs/elemental.yml:56]: will up their damage.

**Comment by brainfood1183** [comment]
@EmoGarbage404 I spiced the supercritical up a little.

https://github.com/space-wizards/space-station-14/assets/113240905/ad3927f1-cc14-4196-87e7-6247c4b15e82

File: Resources/Prototypes/Entities/Mobs/NPCs/elemental.yml (hunk 3 above)
Comment (brainfood1183) [reply Resources/Prototypes/Entities/Mobs/NPCs/elemental.yml]: gave them descriptions.

File: Resources/Prototypes/Entities/Markers/Spawners/Random/crystal.yml (hunk 4 above)
Comment (brainfood1183) [reply Resources/Prototypes/Entities/Markers/Spawners/Random/crystal.yml:6]: suffix added.

File: Resources/Prototypes/Entities/Mobs/NPCs/elemental.yml (hunk 6 above)
Comment (brainfood1183) [reply Resources/Prototypes/Entities/Mobs/NPCs/elemental.yml:56]: damage now 2x what it was before.

File: Resources/Prototypes/Entities/Mobs/NPCs/elemental.yml (hunk 2 above)
Comment (brainfood1183) [reply Resources/Prototypes/Entities/Mobs/NPCs/elemental.yml:4]: I checked all base types and like drones i think elementals need their own base so they dont inherit stuff like flammable and body solutions.

**Comment by CatMagic1** [comment]
Can't wait for the halls to become a cave maze.

File: Content.Shared/Anomaly/Effects/Components/EntitySpawnAnomalyComponent.cs [hunk 7]
```diff
@@ -14,13 +14,19 @@ public sealed partial class EntitySpawnAnomalyComponent : Component
     [DataField("spawns", required: true, customTypeSerializer: typeof(PrototypeIdListSerializer<EntityPrototype>)), ViewVariables(VVAccess.ReadWrite)]
     public List<string> Spawns = new();
 
+    [DataField("superCriticalSpawns", required: true, customTypeSerializer: typeof(PrototypeIdListSerializer<EntityPrototype>)), ViewVariables(VVAccess.ReadWrite)]
+    public List<string> SuperCriticalSpawns = new();
```
Comment (EmoGarbage404) [inline Content.Shared/Anomaly/Effects/Components/EntitySpawnAnomalyComponent.cs:18]: you can just replace this with a list of `EntProtoId`s

File: Content.Server/Anomaly/Effects/EntityAnomalySystem.cs [hunk 8]
```diff
@@ -68,15 +72,22 @@ private void SpawnMonstersOnOpenTiles(EntitySpawnAnomalyComponent component, Tra
...
+                if (component.WallEntity == true && body.BodyType != BodyType.Static)
                     continue;
+
+                if (component.WallEntity == false)
+                {
+                    if (body.BodyType != BodyType.Static ||
+                        !body.Hard ||
+                        (body.CollisionLayer & (int) CollisionGroup.Impassable) == 0)
+                        continue;
+                }
```
Comment (EmoGarbage404) [inline Content.Server/Anomaly/Effects/EntityAnomalySystem.cs:85]: what's the practical difference that requires this?

File: Content.Server/Anomaly/Effects/EntityAnomalySystem.cs [hunk 9]
```diff
@@ -68,15 +72,22 @@ private void SpawnMonstersOnOpenTiles(EntitySpawnAnomalyComponent component, Tra
...
                 valid = false;
                 break;
             }
             if (!valid)
-                continue;
+            continue;
```
Comment (EmoGarbage404) [inline Content.Server/Anomaly/Effects/EntityAnomalySystem.cs:90]: ate the indent

**Review by EmoGarbage404 — CHANGES_REQUESTED** [review]

File: Content.Server/Anomaly/Effects/EntityAnomalySystem.cs (hunk 8 above)
Comment (brainfood1183) [reply Content.Server/Anomaly/Effects/EntityAnomalySystem.cs:85]: stops walls spawning on tables and in doorways, not a big issue gameplay wise but looks like ass

File: Content.Server/Anomaly/Effects/EntityAnomalySystem.cs [hunk 10]
```diff
@@ -9,13 +9,16 @@
...
+using Content.Shared.Mobs.Components;
+using Microsoft.EntityFrameworkCore.Metadata.Builders;
 
 namespace Content.Server.Anomaly.Effects;
 
 public sealed class EntityAnomalySystem : EntitySystem
 {
     [Dependency] private readonly IMapManager _map = default!;
     [Dependency] private readonly IRobustRandom _random = default!;
+    [Dependency] private readonly EntityLookupSystem _lookup = default!;
```
[the remark this answers is not in the document]
Comment (brainfood1183) [reply Content.Server/Anomaly/Effects/EntityAnomalySystem.cs:21]: removed it.

File: Content.Shared/Anomaly/Effects/Components/EntitySpawnAnomalyComponent.cs [hunk 11]
```diff
@@ -14,13 +14,19 @@ public sealed partial class EntitySpawnAnomalyComponent : Component
...
+
     /// <summary>
     /// The maximum number of entities that spawn per pulse
     /// scales with severity.
     /// </summary>
     [DataField("maxSpawnAmount"), ViewVariables(VVAccess.ReadWrite)]
     public int MaxSpawnAmount = 7;
 
+    [DataField("wallEntity")]
+    public bool WallEntity = false;
```
[the remark this answers is not in the document]
Comment (brainfood1183) [reply Content.Shared/Anomaly/Effects/Components/EntitySpawnAnomalyComponent.cs:28]: added documentation and removed the name inferal.

File: Content.Shared/Anomaly/Effects/Components/EntitySpawnAnomalyComponent.cs (hunk 7 above)
Comment (brainfood1183) [reply Content.Shared/Anomaly/Effects/Components/EntitySpawnAnomalyComponent.cs:18]: No idea how to use EntProtoId and have it tie in with the way the entities are spawned in the system.

File: Content.Server/Anomaly/Effects/EntityAnomalySystem.cs (hunk 9 above)
Comment (brainfood1183) [reply Content.Server/Anomaly/Effects/EntityAnomalySystem.cs:90]: indent has be regurgitated.

File: Content.Shared/Anomaly/Effects/Components/TileSpawnAnomaly.cs [hunk 12]
```diff
@@ -14,6 +14,9 @@ public sealed partial class TileSpawnAnomalyComponent : Component
     [DataField("spawnRange"), ViewVariables(VVAccess.ReadWrite)]
     public float SpawnRange = 5f;
 
+    [DataField("spawnChance"), ViewVariables(VVAccess.ReadWrite)]
+    public float SpawnChance = 0.33f;
```
[the remark this answers is not in the document]
Comment (brainfood1183) [reply Content.Shared/Anomaly/Effects/Components/TileSpawnAnomaly.cs]: done.

File: Content.Shared/Anomaly/Effects/Components/EntitySpawnAnomalyComponent.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (brainfood1183) [reply Content.Shared/Anomaly/Effects/Components/EntitySpawnAnomalyComponent.cs]: decided to remove this in the end.

File: Content.Server/Anomaly/Effects/EntityAnomalySystem.cs (hunk 8 above)
Comment (brainfood1183) [reply Content.Server/Anomaly/Effects/EntityAnomalySystem.cs]: I decided to remove this bool in the end, it causes other issues and to get it perfect would require a very lengthy if statement.

File: Content.Shared/Anomaly/Effects/Components/EntitySpawnAnomalyComponent.cs (hunk 7 above)
Comment (brainfood1183) [reply Content.Shared/Anomaly/Effects/Components/EntitySpawnAnomalyComponent.cs]: this has been done.

File: Resources/Prototypes/Entities/Mobs/NPCs/elemental.yml [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+        damage: 40
+      behaviors:
+      - !type:PlaySoundBehavior
+        sound:
+          collection: GlassBreak
+      - !type:SpawnEntitiesBehavior
+        spawn:
+          SpaceQuartz:
+            min: 1
+            max: 2
```
Comment (mirrorcult) [inline Resources/Prototypes/Entities/Mobs/NPCs/elemental.yml:114]: the crabs should spawn more ore than this, maybe 4-6

**Review by mirrorcult — CHANGES_REQUESTED** [review]

File: Resources/Prototypes/Entities/Mobs/NPCs/elemental.yml (hunk 13 above)
Comment (brainfood1183) [reply Resources/Prototypes/Entities/Mobs/NPCs/elemental.yml:114]: sure i can increase it.

**Review by mirrorcult — APPROVED** [review]

**Comment by mirrorcult** [comment]
more....... anobblies...............

**Pull request merged**